Newspack video management proposal & Video Playlist block

claudiulodro opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-theme/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

This PR is an interactive proposal for video management under Newspack. YouTube is the most common platform our publishers use, and Gutenberg has an excellent YouTube block. This PR adds a Video Playlist block. Together, these two blocks can be a complete video management and display solution.

- For single videos, embed a YouTube block.
- For featured video areas (Latest videos, a video series, etc.) use the Video Playlist block.
- The Video Playlist block can either automatically pull the latest videos from among the YouTube videos embedded in posts or can build a custom playlist of manually selected videos.
- With a block-based approach, the need for centralized video management is eliminated.
- If you want to have a video or video playlist "featured image", select the Hidden featured image style and put a video at the top of the post. It will look just like a video featured image.

### How to test the changes in this Pull Request:

We can probably hold off on code review until we've figured out the right feature set and other implementation details and are thinking about merging this.

1. Check out and build this branch.
2. Add YouTube embeds to a variety of published posts:
<img width="679" alt="Screen Shot 2020-02-07 at 10 59 36 AM" src="https://user-images.githubusercontent.com/7317227/74058488-04f2c900-499b-11ea-88d6-a85358c8e8ce.png">
3. On a new post, insert the Video Playlist block. It should auto-populate with the videos you've inserted around your site. Try the limit setting also.
<img width="602" alt="Screen Shot 2020-02-07 at 11 01 39 AM" src="https://user-images.githubusercontent.com/7317227/74058645-54d19000-499b-11ea-9e55-ca21f3fd63dc.png">
4. Test the Category filtering option by selecting a category. Only videos from posts in that category should be in the playlist.
<img width="604" alt="Screen Shot 2020-02-07 at 11 02 09 AM" src="https://user-images.githubusercontent.com/7317227/74058673-6155e880-499b-11ea-9628-46c1ffa9524f.png">
5. Test the manual mode.
<img width="678" alt="Screen Shot 2020-02-07 at 11 06 37 AM" src="https://user-images.githubusercontent.com/7317227/74058699-6c107d80-499b-11ea-9591-f10f7f839e98.png">
6. Test the Autoplay toggle. It should autoplay the video on the frontend.

### Discussion/Questions

1. On the frontend, the video playlist is sort-of hidden and you need to click the hamburger menu in the upper-right corner to display it. I don't think most people will notice that this is a playlist because of that. There is not a way to force the embed to start with that panel open as far as I can tell. It might need a custom player. Additionally, the title is always 'Untitled Playlist', which is not something that appears change-able.
<img width="721" alt="Screen Shot 2020-02-07 at 11 19 19 AM" src="https://user-images.githubusercontent.com/7317227/74058881-cf021480-499b-11ea-83b6-df9956e28c08.png">
<img width="716" alt="Screen Shot 2020-02-07 at 11 19 26 AM" src="https://user-images.githubusercontent.com/7317227/74058889-d2959b80-499b-11ea-9195-2a933c8ab2b9.png">

There are a few different approaches we could take if it turns out we need to build a custom player, but the current way the embeds are done in this PR is the simplest way, so I'm starting with it to get the discussion going.

2. This block solely supports YouTube videos. Should it be renamed to something like "YouTube Playlist"? If we support other providers in the future, will they be part of this block? Will they have their own blocks (e.g. "JWPlayer Playlist")?

3. Is that TokenField the way to go for manual YouTube playlist building? Is there a better UI/UX for that?

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

cc @deke207

thomasguillot commented on src/blocks/video-playlist/index.js:
```diff
@@ -0,0 +1,69 @@
+export const title = __( 'Video Playlist', 'newspack-blocks' );
+
+/* From https://material.io/tools/icons */
+export const icon = (
```
This icon feels a bit abstract. Let's use "subscription" instead: https://material.io/resources/icons/?icon=subscriptions&style=baseline

claudiulodro replied:
Much nicer. https://github.com/Automattic/newspack-blocks/pull/362/commits/48ad241a1ed5b0ca23be5695681752e9c76f4705

thomasguillot commented on src/blocks/video-playlist/index.js:
```diff
@@ -0,0 +1,69 @@
+	},
+	supports: {
+		html: false,
+		align: false,
```
I think we should support `align`, just like the YouTube block

claudiulodro replied:
Much nicer. https://github.com/Automattic/newspack-blocks/pull/362/commits/48ad241a1ed5b0ca23be5695681752e9c76f4705

claudiulodro replied:
https://github.com/Automattic/newspack-blocks/pull/362/commits/11c54f26b695a5d390a7ce29cb21297286173a9c

thomasguillot commented on src/blocks/video-playlist/edit.js:
```diff
@@ -0,0 +1,270 @@
+	/**
+	 * Add a manual YouTube video from the field.
+	 *
+	 * @param {string} tokens Video URL from the tokenfield.
```
I'm not a huge fan of using `FormTokenField` for the manual URLs, it's an odd UI for URLs and should only be used for categories/tags.
But there aren't many Core alternatives...
The only one I can think of is combining text+button
![Screenshot 2020-02-10 at 15 23 09](https://user-images.githubusercontent.com/177929/74163120-cf0f4800-4c19-11ea-8861-1b892c705bcc.png)
And always allowing an extra empty row for the user to add more videos.

claudiulodro commented:
**In the interest of getting something out into the world, I've removed the UI for the manually selected videos from this PR. Let's discuss and handle that in a follow-up PR.**

> In the editor I'm not able to interact with the YouTube embed. By contrast I can still interact with the core YouTube block, so there should be a good path to resolving.

Makes sense. I've added this in https://github.com/Automattic/newspack-blocks/pull/362/commits/d1dce97dcb33233ff5ef9e3ba612e2238c680b58

> I wonder if the block should have YouTube in the name, to be very clear that this covers YouTube videos only? If we decided we wanted to support other playlists — uploaded videos, Vimeo, etc — presumably these would be separate blocks?

Good call. I think those would be separate blocks also. https://github.com/Automattic/newspack-blocks/pull/362/commits/aa0beda431aca6761671af07ef9cc9f14d7e667d

> Would there ever be a need for a paginated version of this feature? I'm wondering about a case where a publication has years of posts, perhaps with hundreds of videos. There would be no way to curate beyond the most recent thirty except by using categories.

I'm not sure about this use-case. We can re-approach in a future release if a publisher absolutely needs something like that. If you scroll down the homepage towards the bottom video sections of the homepages at [BDN](https://bangordailynews.com/) and [Kawowo](https://kawowo.com/) you can see they have recent video sections. This feature is supposed to enable those.

jeffersonrabb commented on src/blocks/video-playlist/view.js:
```diff
@@ -0,0 +1,5 @@
+/**
```
Can we delete this file entirely? In most cases we use it only to import `view.scss` but in this case that's empty too.

claudiulodro replied:
Yeah good call. https://github.com/Automattic/newspack-blocks/pull/362/commits/790842f28f04dfcdab7f9bce46497b97fd9b8869

jeffersonrabb commented on src/blocks/video-playlist/view.scss:
```diff
@@ -0,0 +1,3 @@
+/**
```
This file can be deleted.

claudiulodro replied:
https://github.com/Automattic/newspack-blocks/pull/362/commits/790842f28f04dfcdab7f9bce46497b97fd9b8869

jeffersonrabb commented on src/blocks/video-playlist/index.js:
```diff
@@ -0,0 +1,69 @@
+ * Style dependencies - will load in editor
+ */
+import './editor.scss';
+import './view.scss';
```
Can be removed.

claudiulodro replied:
https://github.com/Automattic/newspack-blocks/pull/362/commits/790842f28f04dfcdab7f9bce46497b97fd9b8869

jeffersonrabb commented on src/blocks/video-playlist/edit.js:
```diff
@@ -0,0 +1,269 @@
+				<InspectorControls>
+					<PanelBody title={ __( 'Settings', 'newspack-blocks' ) } initialOpen={ true }>
+						<PanelRow>
+							<ToggleControl
```
Autoplay doesn't work on mobile or AMP, so for most Newspack customers this feature will appear to be broken. What do you think about removing?

claudiulodro replied:
It works on AMP in my testing, but we should test more and add that warning for mobile if we decide to keep the feature. I've removed it in https://github.com/Automattic/newspack-blocks/pull/362/commits/11396d82f5812980240dcdc294ad1281efe79b74 until we figure this out further.

jeffersonrabb commented on src/blocks/video-playlist/index.js:
```diff
@@ -0,0 +1,69 @@
+		manual: {
+			type: 'boolean',
+		},
+		videos: {
```
This is the attribute we used for picking specific YouTube videos, correct? Should we remove it to avoid confusion, or would you prefer to leave it in since that feature will be revisited in a separate PR?

claudiulodro replied:
As a best-practice, let's remove it. It's straightforward to add it back since we have the commit history. https://github.com/Automattic/newspack-blocks/pull/362/commits/f167d20b18ddb27b2146f45933c15ab13fb653ae

jeffersonrabb commented on src/blocks/video-playlist/index.js:
```diff
@@ -0,0 +1,69 @@
+		className: {
+			type: 'string',
+		},
+		manual: {
```
Remove reference to `manual`?

claudiulodro replied:
https://github.com/Automattic/newspack-blocks/pull/362/commits/f167d20b18ddb27b2146f45933c15ab13fb653ae

jeffersonrabb commented on src/blocks/video-playlist/view.php:
```diff
@@ -0,0 +1,202 @@
+	$videos = array();
+	$posts  = get_posts( $query_args );
+	foreach ( $posts as $post ) {
+		$blocks         = parse_blocks( $post->post_content );
```
This is cool!
